drivers/charger/bq24190.c

```diff
@@ -0,0 +1,485 @@
+		return ret;
+	}
+
+	v = FIELD_GET(BQ24190_REG_PCTCC_IPRECHG_MASK, v);
+
+	pc = (v * BQ24190_REG_PCTCC_IPRECHG_STEP_UA) + BQ24190_REG_PCTCC_IPRECHG_OFFSET_UA;
+
+	if (frc_20pct) {
+		pc /= 2;
+	}
+
+	val->precharge_curr = pc;
+
+	return 0;
+}
+
+static int bq24190_charger_get_tc(const struct device *dev, union charger_propval *val)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint32_t tc;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_PCTCC, &v);
+	if (ret) {
+		return ret;
+	}
+
+	v = FIELD_GET(BQ24190_REG_PCTCC_ITERM_MASK, v);
+
+	tc = (v * BQ24190_REG_PCTCC_ITERM_STEP_UA) + BQ24190_REG_PCTCC_ITERM_OFFSET_UA;
+
+	val->charge_term_curr = tc;
+
+	return 0;
+}
+
+static int bq24190_get_cvc(const struct device *dev, union charger_propval *val)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint32_t voltage;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_CVC, &v);
+	if (ret < 0) {
+		return ret;
+	}
+
+	v = FIELD_GET(BQ24190_REG_CVC_VREG_MASK, v);
+
+	voltage = (v * BQ24190_REG_CVC_VREG_STEP_UV) + BQ24190_REG_CVC_VREG_OFFSET_UV;
+
+	val->const_charge_volt = voltage;
+
+	return 0;
+}
+
+static int bq24190_set_ccc(const struct device *dev, const union charger_propval *val)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint32_t curr = val->const_charge_curr;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_CCC, &v);
+	if (ret < 0) {
+		return ret;
+	}
+
+	v &= BQ24190_REG_CCC_FORCE_20PCT_MASK;
+
+	if (v) {
+		curr *= 5;
+	}
+
+	curr = CLAMP(curr, BQ24190_REG_CCC_ICHG_MIN_UA, BQ24190_REG_CCC_ICHG_MAX_UA);
+
+	v = (curr - BQ24190_REG_CCC_ICHG_OFFSET_UA) / BQ24190_REG_CCC_ICHG_STEP_UA;
+
+	v = FIELD_PREP(BQ24190_REG_CCC_ICHG_MASK, v);
+
+	return i2c_reg_update_byte_dt(&config->i2c, BQ24190_REG_CCC, BQ24190_REG_CCC_ICHG_MASK, v);
+}
+
+static int bq24190_set_cvc(const struct device *dev, const union charger_propval *val)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint32_t voltage = val->const_charge_volt;
+	uint8_t v;
+
+	voltage = CLAMP(voltage, BQ24190_REG_CVC_VREG_MIN_UV, BQ24190_REG_CVC_VREG_MAX_UV);
+
+	v = (voltage - BQ24190_REG_CVC_VREG_OFFSET_UV) / BQ24190_REG_CVC_VREG_STEP_UV;
+
+	v = FIELD_PREP(BQ24190_REG_CVC_VREG_MASK, v);
+
+	return i2c_reg_update_byte_dt(&config->i2c, BQ24190_REG_CVC, BQ24190_REG_CVC_VREG_MASK, v);
+}
+
+static int bq24190_set_config(const struct device *dev)
+{
+	struct bq24190_data *data = dev->data;
+	union charger_propval val;
+	int ret;
+
+	val.const_charge_curr = data->ichg_ua;
+
+	ret = bq24190_set_ccc(dev, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val.const_charge_volt = data->vreg_uv;
+
+	return bq24190_set_cvc(dev, &val);
+}
+
+static int bq24190_get_prop(const struct device *dev, charger_prop_t prop,
+			    union charger_propval *val)
+{
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		return bq24190_charger_get_online(dev, val);
```

Comment:
I don't think it make much sense to pass `val` in the union form to these functions, this should rather be a

```c
case CHARGER_PROP_ONLINE:
	return bq24190_charger_get_online(dev, &val->online);
```

Then in code review you can imediately see that the union has been resolved to the right type form the fact that `CHARGER_PROP_ONLINE` and `online` are next to each other and there's no way you can pick the wrong one in the function implementation.

Reply:
ack